src/Migrations/HSLMigration.php

```diff
@@ -0,0 +1,646 @@
+      }
+    } elseif ($fn_name === 'Str\\slice' && C\count($items) === 3) {
+      // check for negative length arguments to Str\slice, which will throw a runtime exception
+      $length = $this->resolveIntegerArgument($items[2]);
+      if ($length !== null && $length < 0) {
+        $offset = $this->resolveIntegerArgument($items[1]);
+        if ($offset === null) {
+          // skip this one if we don't have a sensible offset
+          return null;
+        }
+
+        // if the offset is negative too, it's pretty simple
+        // we can compute the correct length as abs(offset) + length and rewrite teh node
+        if ($offset < 0) {
+          $rewrite_length_value = Math\abs($offset) + $length;
+          $unary = C\onlyx(
+            $items[2]->getDescendantsOfType(PrefixUnaryExpression::class),
+          );
+          $new_length = new ListItem(
+            new LiteralExpression(new DecimalLiteralToken(
+              HHAST\Missing(),
+              HHAST\Missing(),
+              (string)$rewrite_length_value,
+            )),
+            HHAST\Missing(),
+          );
+        } else {
+          // with a positive offset this is harder
+          // we need to replace this arg with a more complex expression
+          // based on the length of the string
+          $haystack = $items[0]->getCode();
+          $new_length = $this->nodeFromCode(
+            'Str\\length('.$haystack.') - '.($offset + Math\abs($length)),
+            ExpressionStatement::class,
+          );
+        }
+
+        // rewrite args list
+        $new_argument_list = $argument_list->replace($items[2], $new_length);
+      }
+    } elseif ($fn_name === 'Str\\splice' && C\count($items) === 4) {
+      // check for negative length arguments to Str\splice, which will throw a runtime exception
+      // this is currently unhandled, so we just bail by returning null if we find it
+      $length = $this->resolveIntegerArgument($items[3]);
+      if ($length !== null && $length < 0) {
+        return null;
+      }
+    } elseif (
+      ($fn_name === 'Math\\max' || $fn_name === 'Math\\min') &&
+      C\count($items) !== 1
+    ) {
+      // PHP max() and min() either take a list of variadic args, or an array of args
+      // in HSL, max and min want a single Traversable arg, while maxva and minva are variadic
+      return $this->replaceFunctionName($node, $fn_name.'va');
+    } elseif ($fn_name === 'Math\\round' && C\count($items) > 2) {
+      // can't handle the optional third argument of round()
+      return null;
+    }
+
+    if ($argument_order !== null) {
+
+      $new_items = vec[];
+      foreach ($argument_order as $index) {
+        $new_items[] = $items[$index];
+      }
+
+      $new_argument_list = vec[];
+
+      foreach ($arguments as $i => $argument) {
+        invariant($argument instanceof ListItem, 'expected ListItem');
+        $new_argument_list[] =
+          $argument->replace($argument->getItem(), $new_items[(int)$i]);
```

Comment:
Is there a test covering this, both with and without trailing commas?

e.g, we can't turn `foo(1, 2, 3)` into `bar(3 1, 2,)` ?

Reply:
Added a test for this (the commas stay exactly where they are)